src/passes/Print.cpp

```diff
@@ -141,37 +88,112 @@ struct TypeName {
   TypeName(Type type) : type(type) {}
 };
 
-struct HeapTypeName {
-  HeapType type;
-  HeapTypeName(HeapType type) : type(type) {}
+struct ResultTypeName {
+  Type type;
+  ResultTypeName(Type type) : type(type) {}
 };
 
+static void
+printHeapTypeName(std::ostream& os, HeapType type, bool first = true);
+
+static void printTypeName(std::ostream& os, Type type) {
+  if (type.isBasic()) {
+    os << type;
+    return;
+  }
+  if (type.isTuple()) {
+    auto sep = "";
+    for (auto t : type) {
+      os << sep;
+      sep = "_";
+      printTypeName(os, t);
+    }
+  } else if (type.isRef()) {
+    os << "ref";
+    if (type.isNullable()) {
+      os << "?";
+    }
+    os << "{";
```

Comment:
I think this used to be `[`? Any reason to change?

Reply:
Oh, I just happened to notice that was ambiguous, since we use `"[ ]"` for Arrays too (and it seems to fit Arrays more).

Reply:
But don't we use `{}` for structs instead?

Reply:
Oh right... how silly of me :rofl: 

Then maybe `< >`..? Or something else?

Reply:
It might be harder to read if they are mixed with `->`s and `=>`s, but other than `()` we don't have any option left, so I'm OK either way..?

Reply:
Yes, the arrows make it hard...

We can do like bash, `ref...fer` :wink:

Or maybe `ref|...|`?

Anyhow, I guess this is not actually ambiguous (since the `ref` prefix disambiguates) so it's not urgent to figure this out.

Reply:
Let's go with `ref|...|` for now, this greatly improves readability over something more ambiguous, e.g.
```wat
-  (local $2 (ref null ${mut:i64_ref?{{i32_f32_f64}}_mut:ref?{{i32_f32_f64}}}))
+  (local $2 (ref null ${mut:i64_ref?|{i32_f32_f64}|_mut:ref?|{i32_f32_f64}|}))
```
(note how the `{{ }}` was ambiguous before)